File: simulations/Design_C.m
classdef Design_C
    % Simulate using essentially Kock & Callot [2015] Experimental Design D
    % (which we here renamed) and retrieve coefficient matrix.
    
    properties
    end

    methods (Static)
        function Yminus0ton = sim_data(n,p,seed,r)
            % INPUTS:
            %   n:          Effective sample size, positive integer
            %   p:          System dimension, positive integer
            %   seed:       Seed for rng
            %   r:          MC iter, integer
            % OUTPUT:
            %   Yminus0ton: (1+n) x p outcome matrix 
            if nargin < 2 || isempty(n) || isempty(p)
                error('Not enough input arguments.');
            end
            if nargin < 4 || isempty(seed) || isempty(r)
                rng('default');
            else
                rng(seed+r);
            end        
            sigma_eps=.1;                             % std(eps_0i) [common]
            nburn=10000;                              % burn-in period
            Yinit=zeros(p,1);
            Ylong=nan(p,nburn+1+n);
            Ylong(:,1)=Yinit;
            Eps=sigma_eps*randn(p,nburn+1+n);         % indep. eps_ti~N(0,sigma_eps^2)
            Theta=Design_C.coef_matrix(p);
            for t=2:nburn+1+n
                Ylong(:,t)=Theta*Ylong(:,t-1)+Eps(:,t);
            end
            Yminus0ton=Ylong(:,nburn+1:nburn+1+n)';   % return as (1+n) x p matrix 
        end

        function Theta = coef_matrix(p)
            Theta=nan(p,p);
            for i=1:p
                for j=1:p
                    Theta(i,j)=((-1)^(abs(i-j)))*(.4^(abs(i-j)+1));
                end
            end
        end
    end
end

File: simulations/design_C.R
# Simulate using essentially Kock & Callot [2015] Experimental Design D (which
# we here renamed) and retrieve coefficient matrix.

## IMPORT COEFFICIENT MATRIX FUNCTION
source("simulations/coef_mats.R") # for coef_mat_c
# coef_mat_c <- function(p) {0}
# insertSource("coef_mats.R", functions = "coef_mat_c")

## MAIN FUNCTION

# INPUTS:
#   n:          Effective sample size, positive integer
#   p:          System dimension, positive integer
#   sigma_eps:  Std.dev of eps_0,i (which are indep. gaussian)
#   seed:       Seed for RNG
#   r:          MC iter, integer
#   nburn:      No. of burn-in periods, integer
# OUTPUT:
#   y0ton: (1+n) x p outcome matrix (after burn-in)
sim_data_c <- function(n = 100, p = 4, sigma_eps = 0.1, nburn = 10000) {
  n_tot <- nburn + 1 + n
  yinit <- matrix(0, p, 1)      # p x 1 initial values (zeros)
  ylong <- matrix(NA, p, n_tot) # p x n_tot outcome matrix
  ylong[, 1] <- yinit           # initiate from all zeros
  normals <- matrix(rnorm(p * n_tot), p, n_tot) # p x n_tot std. normals
  eps <- sigma_eps * normals    # p x n_tot indep. innovations
  # Note: first epsilon not actually in use: kept for simple indexing.
  theta <- coef_mat_c(p)
  for (t in 2:n_tot) {
    ylong[, t] <- as.matrix(theta %*% ylong[, t - 1] + eps[, t])
  }
  # Note: as.matrix used to allow storage
  y0ton <- t(ylong[, (nburn + 1):n_tot])
  return(y0ton)
}